File: scripts/validate_all.py
#!/usr/bin/env python3

"""
validate_all.py

Python validation script for Beltic credentials.
Validates all agent and developer credentials using jsonschema.
"""

import json
import sys
import glob
from pathlib import Path
from typing import List, Tuple
from jsonschema import Draft202012Validator, ValidationError

# Colors for output
class Colors:
    GREEN = '\033[92m'
    RED = '\033[91m'
    YELLOW = '\033[93m'
    BLUE = '\033[94m'
    CYAN = '\033[96m'
    RESET = '\033[0m'

def log(color: str, message: str):
    """Print colored log message."""
    print(f"{color}{message}{Colors.RESET}")

def validate_credential_set(
    schema_path: str,
    pattern: str,
    credential_type: str
) -> Tuple[int, int]:
    """
    Validate a set of credentials against a schema.

    Returns:
        Tuple of (passed, failed) counts
    """
    # Load schema
    try:
        with open(schema_path, 'r') as f:
            schema = json.load(f)
    except FileNotFoundError:
        log(Colors.RED, f"✗ Schema not found: {schema_path}")
        return (0, 1)
    except json.JSONDecodeError as e:
        log(Colors.RED, f"✗ Invalid JSON in schema: {e}")
        return (0, 1)

    # Create validator
    validator = Draft202012Validator(schema)

    # Find credential files
    files = glob.glob(pattern)

    if not files:
        log(Colors.YELLOW, f"  ⚠ No files found matching {pattern}")
        return (0, 0)

    log(Colors.BLUE, f"  Found {len(files)} {credential_type} credential(s)\n")

    passed = 0
    failed = 0

    for file_path in sorted(files):
        file_name = Path(file_path).name
        print(f"    {file_name}... ", end='', flush=True)

        try:
            # Load credential
            with open(file_path, 'r') as f:
                credential = json.load(f)

            # Validate
            errors = list(validator.iter_errors(credential))

            if not errors:
                log(Colors.GREEN, "✓")
                passed += 1
            else:
                log(Colors.RED, "✗")
                print("      Errors:")
                for error in errors[:5]:  # Show first 5 errors
                    path = '/'.join(str(p) for p in error.path) or '(root)'
                    print(f"        - {path}: {error.message}")
                if len(errors) > 5:
                    print(f"        ... and {len(errors) - 5} more")
                failed += 1

        except json.JSONDecodeError as e:
            log(Colors.RED, f"✗ Invalid JSON: {e}")
            failed += 1
        except Exception as e:
            log(Colors.RED, f"✗ Error: {e}")
            failed += 1

    return (passed, failed)

def main():
    """Main validation function."""
    print(f"{Colors.CYAN}╔════════════════════════════════════════════════════════════╗{Colors.RESET}")
    print(f"{Colors.CYAN}║  Beltic Credential Validation - All Credentials           ║{Colors.RESET}")
    print(f"{Colors.CYAN}╚════════════════════════════════════════════════════════════╝{Colors.RESET}\n")

    total_passed = 0
    total_failed = 0

    # Validate Agent Credentials
    log(Colors.CYAN, "═" * 60)
    log(Colors.CYAN, "  Agent Credentials")
    log(Colors.CYAN, "═" * 60)

    agent_passed, agent_failed = validate_credential_set(
        'schemas/agent/v1/agent-credential-v1.schema.json',
        'examples/agent/v1/*.json',
        'agent'
    )

    total_passed += agent_passed
    total_failed += agent_failed

    print(f"  Result: {agent_passed} passed, {agent_failed} failed\n")

    # Validate Developer Credentials (Valid)
    log(Colors.CYAN, "═" * 60)
    log(Colors.CYAN, "  Developer Credentials (Valid Examples)")
    log(Colors.CYAN, "═" * 60)

    dev_passed, dev_failed = validate_credential_set(
        'schemas/developer/v1/developer-credential-v1.schema.json',
        'examples/developer/v1/tests/valid-*.json',
        'developer (valid)'
    )

    total_passed += dev_passed
    total_failed += dev_failed

    print(f"  Result: {dev_passed} passed, {dev_failed} failed\n")

    # Final Summary
    print(f"{Colors.CYAN}╔════════════════════════════════════════════════════════════╗{Colors.RESET}")
    print(f"{Colors.CYAN}║  Validation Summary                                        ║{Colors.RESET}")
    print(f"{Colors.CYAN}╚════════════════════════════════════════════════════════════╝{Colors.RESET}")
    print()
    print(f"  Total Credentials: {total_passed + total_failed}")
    log(Colors.GREEN, f"  Passed:            {total_passed}")

    if total_failed > 0:
        log(Colors.RED, f"  Failed:            {total_failed}")
        print()
        log(Colors.RED, f"✗ {total_failed} credential(s) failed validation")
        sys.exit(1)
    else:
        print()
        log(Colors.GREEN, f"✓ All {total_passed} credentials are valid")
        print()
        sys.exit(0)

if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        print("\n\nValidation interrupted by user")
        sys.exit(130)
    except Exception as e:
        log(Colors.RED, f"\n✗ Fatal error: {e}")
        import traceback
        traceback.print_exc()
        sys.exit(1)

File: scripts/validate-all.js
#!/usr/bin/env node

/**
 * validate-all.js
 *
 * Validates all credentials in the repository (agent + developer, v1 + v2).
 * Used for CI/CD and comprehensive validation checks.
 */

const Ajv = require("ajv/dist/2020");
const addFormats = require("ajv-formats");
const fs = require("fs");
const path = require("path");
const { glob } = require("glob");

const colors = {
  green: '\x1b[32m',
  red: '\x1b[31m',
  yellow: '\x1b[33m',
  blue: '\x1b[34m',
  cyan: '\x1b[36m',
  magenta: '\x1b[35m',
  reset: '\x1b[0m'
};

function log(color, message) {
  console.log(`${colors[color]}${message}${colors.reset}`);
}

async function validateCredentialSet(schemaPath, pattern, credentialType) {
  const ajv = new Ajv({ allErrors: true, strict: false });
  addFormats(ajv);

  // Load schema
  if (!fs.existsSync(schemaPath)) {
    log('red', `✗ Schema not found: ${schemaPath}`);
    return { passed: 0, failed: 1 };
  }

  const schema = JSON.parse(fs.readFileSync(schemaPath, 'utf8'));
  const validate = ajv.compile(schema);

  // Find files
  const files = glob.sync(pattern);

  if (files.length === 0) {
    log('yellow', `  ⚠ No files found matching ${pattern}`);
    return { passed: 0, failed: 0 };
  }

  log('blue', `  Found ${files.length} ${credentialType} credential(s)\n`);

  let passed = 0;
  let failed = 0;

  for (const filePath of files) {
    const fileName = path.basename(filePath);
    process.stdout.write(`    ${fileName}... `);

    try {
      const credential = JSON.parse(fs.readFileSync(filePath, 'utf8'));
      const isValid = validate(credential);

      if (isValid) {
        log('green', '✓');
        passed++;
      } else {
        log('red', '✗');
        console.log(`      Errors:`);
        validate.errors.slice(0, 5).forEach(err => {
          console.log(`        - ${err.instancePath || '(root)'}: ${err.message}`);
        });
        if (validate.errors.length > 5) {
          console.log(`        ... and ${validate.errors.length - 5} more`);
        }
        failed++;
      }
    } catch (error) {
      log('red', `✗ ERROR: ${error.message}`);
      failed++;
    }
  }

  return { passed, failed };
}

async function main() {
  console.log(`${colors.cyan}╔════════════════════════════════════════════════════════════╗${colors.reset}`);
  console.log(`${colors.cyan}║  Beltic Credential Validation - All Credentials           ║${colors.reset}`);
  console.log(`${colors.cyan}╚════════════════════════════════════════════════════════════╝${colors.reset}\n`);

  let totalPassed = 0;
  let totalFailed = 0;

  // ==========================================================================
  // v1 Schemas (Legacy)
  // ==========================================================================

  log('magenta', '══════════════════════════════════════════════════════════════');
  log('magenta', '  SCHEMA VERSION 1 (Legacy)');
  log('magenta', '══════════════════════════════════════════════════════════════\n');

  // Validate Agent Credentials v1
  log('cyan', '═'.repeat(60));
  log('cyan', '  Agent Credentials v1');
  log('cyan', '═'.repeat(60));

  const agentResultsV1 = await validateCredentialSet(
    'schemas/agent/v1/agent-credential-v1.schema.json',
    'examples/agent/v1/valid-*.json',
    'agent v1'
  );

  totalPassed += agentResultsV1.passed;
  totalFailed += agentResultsV1.failed;

  console.log(`  Result: ${agentResultsV1.passed} passed, ${agentResultsV1.failed} failed\n`);

  // Validate Developer Credentials v1 (Valid)
  log('cyan', '═'.repeat(60));
  log('cyan', '  Developer Credentials v1 (Valid Examples)');
  log('cyan', '═'.repeat(60));

  const developerValidResultsV1 = await validateCredentialSet(
    'schemas/developer/v1/developer-credential-v1.schema.json',
    'examples/developer/v1/tests/valid-*.json',
    'developer v1 (valid)'
  );

  totalPassed += developerValidResultsV1.passed;
  totalFailed += developerValidResultsV1.failed;

  console.log(`  Result: ${developerValidResultsV1.passed} passed, ${developerValidResultsV1.failed} failed\n`);

  // ==========================================================================
  // v2 Schemas (W3C VC 2.0 aligned)
  // ==========================================================================

  log('magenta', '══════════════════════════════════════════════════════════════');
  log('magenta', '  SCHEMA VERSION 2 (W3C VC 2.0 aligned)');
  log('magenta', '══════════════════════════════════════════════════════════════\n');

  // Validate Agent Credentials v2
  log('cyan', '═'.repeat(60));
  log('cyan', '  Agent Credentials v2');
  log('cyan', '═'.repeat(60));

  const agentResultsV2 = await validateCredentialSet(
    'schemas/agent/v2/agent-credential-v2.schema.json',
    'examples/agent/v2/valid-*.json',
    'agent v2'
  );

  totalPassed += agentResultsV2.passed;
  totalFailed += agentResultsV2.failed;

  console.log(`  Result: ${agentResultsV2.passed} passed, ${agentResultsV2.failed} failed\n`);

  // Validate Developer Credentials v2 (Valid)
  log('cyan', '═'.repeat(60));
  log('cyan', '  Developer Credentials v2 (Valid Examples)');
  log('cyan', '═'.repeat(60));

  const developerValidResultsV2 = await validateCredentialSet(
    'schemas/developer/v2/developer-credential-v2.schema.json',
    'examples/developer/v2/tests/valid-*.json',
    'developer v2 (valid)'
  );

  totalPassed += developerValidResultsV2.passed;
  totalFailed += developerValidResultsV2.failed;

  console.log(`  Result: ${developerValidResultsV2.passed} passed, ${developerValidResultsV2.failed} failed\n`);

  // ==========================================================================
  // Final Summary
  // ==========================================================================

  console.log(`${colors.cyan}╔════════════════════════════════════════════════════════════╗${colors.reset}`);
  console.log(`${colors.cyan}║  Validation Summary                                        ║${colors.reset}`);
  console.log(`${colors.cyan}╚════════════════════════════════════════════════════════════╝${colors.reset}`);
  console.log('');
  console.log(`  Total Credentials: ${totalPassed + totalFailed}`);
  console.log(`    v1: ${agentResultsV1.passed + developerValidResultsV1.passed} passed, ${agentResultsV1.failed + developerValidResultsV1.failed} failed`);
  console.log(`    v2: ${agentResultsV2.passed + developerValidResultsV2.passed} passed, ${agentResultsV2.failed + developerValidResultsV2.failed} failed`);
  log('green', `  Passed:            ${totalPassed}`);

  if (totalFailed > 0) {
    log('red', `  Failed:            ${totalFailed}`);
    console.log('');
    log('red', `✗ ${totalFailed} credential(s) failed validation`);
    process.exit(1);
  } else {
    console.log('');
    log('green', `✓ All ${totalPassed} credentials are valid`);
    console.log('');
    process.exit(0);
  }
}

main().catch(error => {
  log('red', `\n✗ Fatal error: ${error.message}`);
  console.error(error.stack);
  process.exit(1);
});
